### PR TITLE
adding changes to model2

### PR DESCRIPTION
**Pull request by rohinikapoor**
comments added

### CODE REVIEW

File: code/discriminative.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    we also find a message m' which has not been sent from s1->r1. It can be sent by s1 to some other r.
+    we train(s1,r1,m') with a true label of 0.
+    We use the cross entropy loss to calculate our loss.
+    """
+    def __init__(self, epochs=10):
+        self.epochs = epochs
+        # keeps track of how many times the model has seen each email_id, either as a sender or receiver
+        self.emailid_train_freq = {}
+        super(DiscriminativeModel, self).__init__()
+        #50+50+50 (sender+recr+msg)
```
Comment (mohit-surana) [inline code/discriminative.py:29]: You could modify the comment and the code to use the constants
Refer #32

File: code/discriminative.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        r_emb = self.embedding_layer(r_id)
+        email_emb = email_emb.unsqueeze(0)
+        # simple concatenation of sender and receiver embedding with the email
+        sr_emb = torch.cat((s_emb, r_emb,email_emb), 1)
+        out = self.linear(sr_emb)
+        return out
+
+    def train(self, emails, all_emails, w2v):
+        all_emails_len = len(all_emails)
+        loss_criteria = nn.CrossEntropyLoss()
```
Comment (mohit-surana) [inline code/discriminative.py:53]: I'm curious to know if there is any difference between BCE and CrossEntropy in terms of how it works internally
Perhaps more efficiently?

Will update this comment if I find this to be true.

File: code/discriminative.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+        for epoch in range(self.epochs):
+            print 'running epoch ', epoch
+            start = time.time()
+            epoch_loss = 0.0
+            for i in range(len(emails)):
+                # if i%1000 == 0:
+                #     print 'email processed =', i
+                sender_id = utils.get_userid(emails[i, constants.SENDER_EMAIL])
+                email_word_reps = w2v.get_sentence(emails[i, constants.EMAIL_BODY])
```
Comment (mohit-surana) [inline code/discriminative.py:67]: #33 could be implemented here as well
Either in this PR or a future PR one of us can implement this.

File: code/discriminative.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                    
+                    loss = loss_criteria(pred_out,pos_label)
+                    # propagate the loss backward and compute the gradient
+                    loss.backward()
+                    # change weights based on gradient value
+                    optimizer.step()
+
+                    #now train for a negative sample
+                    #get a random email from all_emails where the receiver is not recr, and the mail is not empty
+                    while True:
```
Comment (mohit-surana) [inline code/discriminative.py:97]: Above comment applies here too.
Ideally, we should sample mails once and then train over multiple epochs

Based on how long this takes to run we can reprioritize this.

**Review by mohit-surana — COMMENTED** [review]
Looks good overall

File: code/discriminative.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+import torch.nn as nn
+import torch.autograd as autograd
+import torch.optim as optim
+import utils
+import numpy as np
+import time
+import constants
+import random
+
+class DiscriminativeModel(nn.Module, Model):
```
Comment (mohit-surana) [inline code/discriminative.py:12]: As a temporary workaround, renaming this to Model4 would allow you to run this via the command line without making changes to ```main.py```
@mudit1990 Do you think this is a good idea?

Comment (mudit1990) [reply code/discriminative.py:12]: actually, I was thinking the other way round ... let's give our models meaningful names
Model 1,2,3,4 will get very confusing for us ... especially if we are supplying command line args
This can also be run through command line without making any changes in main.py. You just have to pass 'DiscriminativeModel' model as model name

File: code/discriminative.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        super(DiscriminativeModel, self).__init__()
+        #50+50+50 (sender+recr+msg)
+        self.linear = nn.Linear(150, 2)
+        # embedding lookup for 150 users each have 50 dimension representation
+        self.embedding_layer = nn.Embedding(150, 50)
+
+    def forward(self, s_id, r_id, email_emb):
+        """
+        Input are 3 autgrad variables representing sender_id and receiver_id. These ids will be used to lookup the
+        embeddings. Does a forward pass and returns the predicted email representation
```
Comment (mudit1990) [inline code/discriminative.py]: I think the docstring isn't correct. It doesn't return a predicted email representation per say. It will return class scores

File: code/discriminative.py (hunk 3 above)
Comment (mudit1990) [reply code/discriminative.py:67]: yep ... these spend things up quite a bit for the other models

File: code/discriminative.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+                    loss = loss_criteria(pred_out,pos_label)
+                    # propagate the loss backward and compute the gradient
+                    loss.backward()
+                    # change weights based on gradient value
+                    optimizer.step()
+
+                    #now train for a negative sample
+                    #get a random email from all_emails where the receiver is not recr, and the mail is not empty
+                    while True:
+                        msg_indx = random.randint(0,all_emails_len-1)
```
Comment (mudit1990) [inline code/discriminative.py]: I think you should have this sampling code as a separate function

File: code/discriminative.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+                        if utils.get_userid(all_emails[msg_indx, constants.RECEIVER_EMAILS]) != recv_id and len(w2v.get_sentence(all_emails[msg_indx, constants.EMAIL_BODY])) != 0:
+                            break
+
+                    neg_email_word_rep = w2v.get_sentence(all_emails[msg_indx, constants.EMAIL_BODY])
```
Comment (mudit1990) [inline code/discriminative.py]: You need to have an optimizer.zero_grad() before training for negative example.
optimizer.zero_grad() zeros out the gradients of the previous run. If you dont do it gradients from the positive runs will get mixed up with gradients from negative run

File: code/model2.py [hunk 9]
```diff
@@ -13,3 +13,132 @@ def load(self, filename):
...
+import torch.nn as nn
+import torch.autograd as autograd
+import torch.optim as optim
+import utils
+import numpy as np
+
+
+class Model2(nn.Module, Model):
+    """
+    Contains the code for Model3.
```
Comment (mudit1990) [inline code/model2.py]: incorrect/stale docstring

File: code/model2.py [hunk 10]
```diff
@@ -13,3 +13,132 @@ def load(self, filename):
...
+        n_emb = n_emb.unsqueeze(0)
+        #concatiante sender, receiver, prev word and next word embeddings 
+        srpn_emb = torch.cat([s_emb, r_emb,p_emb,n_emb], 1)
+        # srpn_emb = torch.cat((srp_emb, n_emb),1)
+        h1 = self.relu(self.h1_layer(srpn_emb))
+        word_reps = self.email_layer(h1)
+        #the predicted word rep
+        return word_reps
+
+    def get_average_rep(self, word_reps):
```
Comment (mudit1990) [inline code/model2.py:80]: this can be done in-line ... the code has been changed in other models as well

File: code/model2.py [hunk 11]
```diff
@@ -13,3 +13,132 @@ def load(self, filename):
...
+        # this should be the size of <sender+receiver representation>
+        self.h1_layer = nn.Linear(200, 500)
+        # ReLU activation used
+        self.relu = nn.ReLU()
+        # final linear layer that outputs the predicted email representation
+        self.email_layer = nn.Linear(500, 50)
+
+    def forward(self, s_id, r_ids,pv_emb, nv_emb):
+        """
+        Input is an integer sender_id and a list of receiver_ids. These ids are first converted into torch Variables and
```
Comment (mudit1990) [inline code/model2.py:51]: again ... incorrect/stale docstring

File: code/discriminative.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        self.emailid_train_freq = {}
+        super(DiscriminativeModel, self).__init__()
+        #50+50+50 (sender+recr+msg)
+        self.linear = nn.Linear(150, 2)
+        # embedding lookup for 150 users each have 50 dimension representation
+        self.embedding_layer = nn.Embedding(150, 50)
+
+    def forward(self, s_id, r_id, email_emb):
+        """
+        Input are 3 autgrad variables representing sender_id and receiver_id. These ids will be used to lookup the
```
Comment (mudit1990) [inline code/discriminative.py:36]: I think the docstring isn't correct. It doesn't return a predicted email representation per say. It will return class scores

File: code/discriminative.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+                    # propagate the loss backward and compute the gradient
+                    loss.backward()
+                    # change weights based on gradient value
+                    optimizer.step()
+
+                    #now train for a negative sample
+                    #get a random email from all_emails where the receiver is not recr, and the mail is not empty
+                    while True:
+                        msg_indx = random.randint(0,all_emails_len-1)
+                        if utils.get_userid(all_emails[msg_indx, constants.RECEIVER_EMAILS]) != recv_id and len(w2v.get_sentence(all_emails[msg_indx, constants.EMAIL_BODY])) != 0:
```
Comment (mudit1990) [inline code/discriminative.py]: I think the docstring isn't correct. It doesn't return a predicted email representation per say. It will return class scores

File: code/discriminative.py (hunk 8 above)
Comment (mudit1990) [inline code/discriminative.py:102]: You need to have an optimizer.zero_grad() line before this. The function zeros out the gradients of the previous run before doing the forward/backward prop again. If this line is not there the gradients of pos and neg runs will get mixed together

File: code/model2.py [hunk 14]
```diff
@@ -13,3 +13,132 @@ def load(self, filename):
...
+import torch.autograd as autograd
+import torch.optim as optim
+import utils
+import numpy as np
+
+
+class Model2(nn.Module, Model):
+    """
+    Contains the code for Model3.
+    Architecture - sr_embedding -> linear_layer -> relu_activation -> linear_layer -> predicted_email_emb -> loss
```
Comment (mudit1990) [inline code/model2.py:29]: incorrect/stale docstring

**Review by mudit1990 — CHANGES_REQUESTED** [review]

File: code/discriminative.py (hunk 1 above)
Comment (rohinikapoor) [reply code/discriminative.py:29]: yes, i had the previous version of the code. will use the constants.

File: code/discriminative.py (hunk 5 above)
Comment (rohinikapoor) [reply code/discriminative.py:12]: I agree about proper names,
but I tried with 'DiscriminativeModel', but it gave key not found error.
Let me try again.

File: code/discriminative.py (hunk 8 above)
Comment (mohit-surana) [reply code/discriminative.py:102]: Good catch.

File: code/discriminative.py (hunk 5 above)
Comment (mudit1990) [reply code/discriminative.py:12]: have you put the import statement in main.py?
like this one?
from model3 import Model3

File: code/discriminative.py (hunk 3 above)
Comment (mohit-surana) [reply code/discriminative.py:67]: Update: This does not scale beyond 30k mails as of now
You can leave it for now

File: code/discriminative.py (hunk 13 above)
Comment (mohit-surana) [inline code/discriminative.py:99]: This should be `recv_id not in all_emails[msg_indx, constants.RECEIVER_EMAILS]`
